### PR TITLE
 Live video object recognition (Bug fixes)

### PR DESCRIPTION
**Pull request by ghost**
Short: I made a feature where the user can use a webcam to record live frames which then can be recognized with the sample image
Detail: The user has the ability to either:
- show a series or images (in a loop)
- show one image (in a loop)
- insert a file path with the command prompt to show the video feed
- insert a file path to show the video feed
- execute other tasks whilst showing the camera feed
- execute other tasks whilst showing and saving the camera feed
- use a function to apply changes to a video frame, execute tasks whilst showing the camera feed
- record, save and show camera feed

Notice message: Travis shoudl not fail this time.

Changes:
- added new set(cv::mat) functionality to image classes
- used FLANN algorithm to compute good matches
- added current time calculations for demo code in main

Fixes:
- Fixed set() function of RootImage
- Fixed startRecording(bool once, bool showRecording) function of VideoController with looping issue
- Fixed missing .vscode files and CmakeList.txt
- Fixed spelling mistake of previous commit (FLINN --> FLANN)

Author: Arsalan Anwari
Team: Scrum team 6
Date: 26-05-18

### CODE REVIEW

File: src/image/image.hpp [hunk 1]
```diff
@@ -24,13 +26,36 @@ class Image
 {
   protected:
     cv::Mat image;
-
+  
   public:
+    FeatureList features;
```
Comment (wvanbreukelen) [inline src/image/image.hpp:31]: Please write documentation for class members.

File: src/image/image.hpp [hunk 2]
```diff
@@ -24,13 +26,36 @@ class Image
...
+
     /**
      * @brief This function should be used for the intialization process of the image container. 
      */
     virtual void init() = 0;
 
+    /**
+     * @brief This function should be used to fill the image container with an image specified from a path
+     * 
+     * @param new_path : char const*
```
Comment (wvanbreukelen) [inline src/image/image.hpp:42]: ```char const*``` should be a textual description.

File: src/image/root_image.cpp [hunk 3]
```diff
@@ -2,22 +2,55 @@
...
+     void RootImage::init(){
         /// <initialize with imread and settings params, Remove temporary code below
-        for (char & c: image.data){
-            c = '1';
+        set(path);
+    }
+
+
+    void RootImage::set(){
+        if (image.data){ image = cv::Mat(); }
```
Comment (wvanbreukelen) [inline src/image/root_image.cpp:21]: I would suggest to write this line in the form of:

```cpp
 if (image.data) image = cv::Mat();
```
OR

```cpp
if (image.data) {
     image = cv::Mat();
}
```

File: src/image/sample_image.cpp [hunk 4]
```diff
@@ -33,15 +31,39 @@ namespace lucidy{
     void SampleImage::set(SampleImage & other){
         /// copy funtion: clear img.data, setup again with new settings
         
-        /// if len(image) > 0: image.release();
-        image = other.image;
+        if (image.data){ image = cv::Mat(); }
```
Comment (wvanbreukelen) [inline src/image/sample_image.cpp:34]: See comment above

File: src/image/sample_image.hpp [hunk 5]
```diff
@@ -18,12 +18,21 @@ namespace lucidy
  * @brief Sample image wrapper class
  * @details Implements Image interface
  */
-class SampleImage : public Image
+
+
+class SampleImage : public Image 
 {
 private:
   const char *path;
```
Comment (wvanbreukelen) [inline src/image/sample_image.hpp:26]: Please document member functions

File: src/main.cpp [hunk 6]
```diff
@@ -1,14 +1,67 @@
-/**
- * @brief 
- * Main entry point for module.
- * @file main.cpp
- * @author  Dylan Rakiman, Arsalan Anwari
- */
+// #include "object_finding/object_finder.hpp"
```
Comment (wvanbreukelen) [inline src/main.cpp:1]: What is the point of this file? It's fully commented.

File: src/object_finding/feature_comparator.cpp [hunk 7]
```diff
@@ -12,13 +13,40 @@ namespace lucidy{
         /// initalize comparator with custom desires, nessecary has been done in constructor
 
     }
+
+    void FeatureComparator::calcMatchList(RootImage & sourceImage, SampleImage & sampleImage){
+        /// compute descriptors of both root and sample image. match them with descriptor matcher.
+        /// calculated matches output is in matchList. return matchList. 
+        detector.detectFeatures(sourceImage);
+        if(sampleImage.features.size() <= 0 ){detector.detectFeatures(sampleImage);}
```
Comment (wvanbreukelen) [inline src/object_finding/feature_comparator.cpp:21]: Use format suggested.

File: src/object_finding/feature_descriptor.cpp [hunk 8]
```diff
@@ -14,14 +14,15 @@ namespace lucidy{
         descriptor->setUpright( settings.calcUpright );
     }
 
-    DescriptorList& FeatureDescriptor::getDescriptorList(Image & sourceImage){
-        /// compute descriptors by using featureList that has been precalculated by detector
-        /// descriptor output will be placed in descriptorList
+    void FeatureDescriptor::computeDescriptors(Image & sourceImage){
+        /// compute descriptors by using image featureList that has been precalculated by detector
```
Comment (wvanbreukelen) [inline src/object_finding/feature_descriptor.cpp:18]: You may add this in the extended Doxygen description (below @brief) within the header file.

File: src/test_log/image_parser_viewer.cpp [hunk 9]
```diff
@@ -0,0 +1,124 @@
+/**
```
Comment (wvanbreukelen) [inline src/test_log/image_parser_viewer.cpp:1]: What is the point of this file?

File: src/test_log/live_video_feed_object_recognition.cpp [hunk 10]
```diff
@@ -0,0 +1,129 @@
+/**
```
Comment (wvanbreukelen) [inline src/test_log/live_video_feed_object_recognition.cpp:1]: Tip: I would like to see some more info about these tests. What do they test, and how? The strings used as test case items are OK, but descriptions will really improve readablity.
As (as far as I know) the wiki does (sadly) not make any notices about the documentation standard for tests, you should consider this as a improvement point.

File: src/test_log/template_code.cpp [hunk 11]
```diff
@@ -5,25 +5,25 @@
  * @author  Dylan Rakiman, Arsalan Anwari
  */
 
-#include <iostream>
+// #include <iostream>
```
Comment (wvanbreukelen) [inline src/test_log/template_code.cpp:8]: Can these comments be removed? If not, maybe move it into a function so users are still able to run this test if they like.

File: src/window/video_controller.hpp [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+ * @brief This variable is used a static constant which decides if recorded video feed should be displayed inside a window
+ */
+const bool SHOW_VIDEO_FEED = true;
+
+/**
+ * @brief VideoController class
+ */
+class VideoController : public Gui{
+private:
+    RootImageList recordFeed; /// This variable contains all the recored footage as a list of root images
```
Comment (wvanbreukelen) [inline src/window/video_controller.hpp:30]: Here, you are documenting your class members, good job!

File: test/test_main.cpp [hunk 13]
```diff
@@ -0,0 +1,118 @@
+#define CATCH_CONFIG_MAIN // This tells Catch to provide a main() - only do this in one cpp file
+#include "catch.hpp"
+
+#include "object_finding/object_finder.hpp"
+
+TEST_CASE("Feature comparator good match test", "AlgorithmTest"){
```
Comment (wvanbreukelen) [inline test/test_main.cpp:6]: See my previous advice for writing tests documentation.

**Review by wvanbreukelen — CHANGES_REQUESTED** [review]
Please view the placed comments. Note: tips are for your team, you may want to improve it if you like to increase code quality.

**Comment by elsavira** [comment]
Closing pull request: Irrelevant for current situation.

**Pull request closed without merging**